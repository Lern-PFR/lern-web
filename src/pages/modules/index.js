import ModuleDetailsPage from './ModuleDetailsPage';

export {
	// eslint-disable-next-line import/prefer-default-export
	ModuleDetailsPage,
};
