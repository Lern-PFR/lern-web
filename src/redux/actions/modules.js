import * as ModulesApi from 'api/modulesApi';

/**
 * @constant
 * @name ActionTypes
 * @description The various action types used to interact with the modules redux state.
 * @type {object}
 *
 * @author Timothée Simon-Franza
 */
export const ActionTypes = {
	FETCH_MODULE_REQUEST: '@MODULES/FETCH_REQUEST',
	FETCH_MODULE_SUCCESS: '@MODULES/FETCH_SUCCESS',
	FETCH_MODULE_FAILURE: '@MODULES/FETCH_FAILURE',

	FETCH_MODULE_LIST_REQUEST: '@MODULES/FETCH_LIST_REQUEST',
	FETCH_MODULE_LIST_SUCCESS: '@MODULES/FETCH_LIST_SUCCESS',
	FETCH_MODULE_LIST_FAILURE: '@MODULES/FETCH_LIST_FAILURE',

	CREATE_MODULE_REQUEST: '@MODULES/CREATE_REQUEST',
	CREATE_MODULE_SUCCESS: '@MODULES/CREATE_SUCCESS',
	CREATE_MODULE_FAILURE: '@MODULES/CREATE_FAILURE',

	UPDATE_MODULE_REQUEST: '@MODULES/UPDATE_REQUEST',
	UPDATE_MODULE_SUCCESS: '@MODULES/UPDATE_SUCCESS',
	UPDATE_MODULE_FAILURE: '@MODULES/UPDATE_FAILURE',

	DELETE_MODULE_REQUEST: '@MODULES/DELETE_REQUEST',
	DELETE_MODULE_SUCCESS: '@MODULES/DELETE_SUCCESS',
	DELETE_MODULE_FAILURE: '@MODULES/DELETE_FAILURE',

	CLEAR_MODULE_LIST: '@MODULES/CLEAR_LIST',
};

// //////////////////////////////////////////////////////// //
// //////////// Specific module fetching actions ////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchModuleRequest
 * @description Action triggered anytime a module fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchModuleRequest = () => ({ type: ActionTypes.FETCH_MODULE_REQUEST });

/**
 * @function
 * @name fetchModuleSuccess
 * @description Action triggered as a result to a successful module fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The module retrieved from the API.
 *
 * @returns {object}
 */
const fetchModuleSuccess = ({ module }) => ({
	type: ActionTypes.FETCH_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name fetchModuleFailure
 * @description Action triggered as a result to a failed module fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchModuleFailure = (error) => ({
	type: ActionTypes.FETCH_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ////////////// Module list fetching actions //////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchModuleListRequest
 * @description Action triggered anytime a module list fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchModuleListRequest = () => ({ type: ActionTypes.FETCH_MODULE_LIST_REQUEST });

/**
 * @function
 * @name fetchModuleListSuccess
 * @description Action triggered as a result to a successful module list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {array}	modules		: The list of modules retrieved from the API.
 * @param {number}	totalCount	: The total amount of modules available for the subject.
 *
 * @returns {object}
 */
const fetchModuleListSuccess = ({ modules, totalCount }) => ({
	type: ActionTypes.FETCH_MODULE_LIST_SUCCESS,
	payload: { modules, totalCount },
});

/**
 * @function
 * @name fetchModuleListFailure
 * @description Action triggered as a result to a failed module list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchModuleListFailure = (error) => ({
	type: ActionTypes.FETCH_MODULE_LIST_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Module creation actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name createModuleRequest
 * @description Action triggered anytime a module creation call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const createModuleRequest = () => ({ type: ActionTypes.CREATE_MODULE_REQUEST });

/**
 * @function
 * @name createModuleSuccess
 * @description Action triggered as a result to a successfult module creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The newly created module.
 *
 * @returns {object}
 */
const createModuleSuccess = ({ module }) => ({
	type: ActionTypes.CREATE_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name createModuleFailure
 * @description Action triggered as a result to a failed module creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const createModuleFailure = (error) => ({
	type: ActionTypes.CREATE_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Module edition actions //////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name updateModuleRequest
 * @description Action triggered anytime a module edition call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const updateModuleRequest = () => ({ type: ActionTypes.UPDATE_MODULE_REQUEST });

/**
 * @function
 * @name updateModuleSuccess
 * @description Action triggered as a result to a successful module edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The updated module object.
 *
 * @returns {object}
 */
const updateModuleSuccess = ({ module }) => ({
	type: ActionTypes.UPDATE_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name updateModuleFailure
 * @description Action triggered as a result to a failed module edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const updateModuleFailure = (error) => ({
	type: ActionTypes.UPDATE_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Module deletion actions //////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name deleteModuleRequest
 * @description Action triggered anytime a module deletion call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const deleteModuleRequest = () => ({ type: ActionTypes.DELETE_MODULE_REQUEST });

/**
 * @function
 * @name deleteModuleSuccess
 * @description Action triggered as a result to a successful module deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The deleted module object.
 *
 * @returns {object}
 */
const deleteModuleSuccess = ({ module }) => ({
	type: ActionTypes.DELETE_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name deleteModuleFailure
 * @description Action triggered as a result to a failed module deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const deleteModuleFailure = (error) => ({
	type: ActionTypes.DELETE_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Exported action creators ////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchModule
 * @description Method used to fetch the latest version of a specified module
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId : The id of the module we want to retrieve.
 */
export const fetchModule = (moduleId) => (dispatch) => {
	dispatch(fetchModuleRequest());

	return ModulesApi.fetchModuleById(moduleId)
		.then(({ module }) => dispatch(fetchModuleSuccess({ module })))
		.catch((error) => dispatch(fetchModuleFailure(error)));
};

/**
 * @function
 * @name fetchModuleListBySubjectId
 * @description Method used to fetch all modules from a specific subject, identified by the subjectId param.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} subjectId : The id of the subject we want to retrieve modules from.
 */
export const fetchModuleListBySubjectId = (subjectId) => (dispatch) => {
	dispatch(fetchModuleListRequest());

	return ModulesApi.fetchModulesBySubjectId(subjectId)
		.then(({ modules, totalCount }) => dispatch(fetchModuleListSuccess({ modules, totalCount })))
		.catch((error) => dispatch(fetchModuleListFailure(error)));
};

/**
 * @function
 * @name createModule
 * @description Method used to create a new module in the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} moduleData : The data to create the new module from.
 */
export const createModule = (moduleData) => (dispatch) => {
	dispatch(createModuleRequest());

	return ModulesApi.createModule(moduleData)
		.then(({ module }) => dispatch(createModuleSuccess({ module })))
		.catch((error) => dispatch(createModuleFailure(error)));
};

/**
 * @function
 * @name updateModule
 * @description Method used to update an existing module instance from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} moduleData	: The data to update the module with.
 * @param {string} moduleId		: The id to identify the module to update.
 */
export const updateModule = (moduleData, moduleId) => (dispatch) => {
	dispatch(updateModuleRequest());

	return ModulesApi.updateModule(moduleData, moduleId)
		.then(({ module }) => dispatch(updateModuleSuccess({ module })))
		.catch((error) => dispatch(updateModuleFailure(error)));
};

/**
 * @function
 * @name deleteModule
 * @description Method used to remove an existing module from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId : The id identifying the module to remove.
 */
export const deleteModule = (moduleId) => (dispatch) => {
	dispatch(deleteModuleRequest());

	return ModulesApi.deleteModule(moduleId)
		.then(({ module }) => dispatch(deleteModuleSuccess({ module })))
		.catch((error) => dispatch(deleteModuleFailure(error)));
};
