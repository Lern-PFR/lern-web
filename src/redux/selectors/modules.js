import { sortBy } from 'lodash';
import { createSelector } from 'reselect';
import { getSubjectById } from './subjects';

/**
 * @function
 * @name getModules
 * @description A selector callback which returns all modules from the Redux state.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} state The redux state at the time of call.
 *
 * @returns {Array}
 */
const getModules = (state) => state?.modules?.items ?? [];

/**
 * @function
 * @name getModuleById
 * @description A selector callback which returns the module whose id matched the moduleId parameter.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId The id of the module we want to retrieve.
 *
 * @returns {object|undefined}
 */
const getModuleById = createSelector(
	getModules,
	(_, moduleId) => moduleId,
	(modules, moduleId) => {
		const module = modules.filter(({ id }) => (id === moduleId))?.[0] ?? undefined;

		if (module === undefined) {
			return module;
		}

		const concepts = (module.concepts ?? []).map((concept) => {
			const { courses, ...result } = concept; // @TODO: remove this line once the API has been updated.

			const lessons = Object.values((concept.courses || []).reduce((acc, currentLesson) => {
				const currentlyStoredLesson = acc[currentLesson.id];
				if (currentlyStoredLesson) {
					acc[currentLesson.id] = currentlyStoredLesson.version < currentLesson.version ? currentLesson : currentlyStoredLesson;
				} else {
					acc[currentLesson.id] = currentLesson;
				}

				return acc;
			}, {}));

			return {
				...result,
				lessons: sortBy(lessons, 'order'),
			};
		});

		return {
			...module,
			concepts: sortBy(concepts || [], 'order'),
		};
	}
);

/**
 * @function
 * @name getModuleOrderOptions
 * @description A selector callback which returns an array of "order" values for the module edition form's order select field.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {Array}
 */
const getModuleOrderOptions = createSelector(
	getSubjectById,
	(subject) => {
		if (!subject?.modules?.length || subject?.modules?.length < 1) {
			return [{ label: 0, value: 0 }];
		}

		return Array.from({ length: subject?.modules?.length || 0 }, (_, i) => ({ label: i, value: i }));
	}
);

export {
	getModules,
	getModuleById,
	getModuleOrderOptions,
};
