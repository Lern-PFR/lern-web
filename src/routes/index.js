import keys from './keys';

export {
	keys as default,
};
