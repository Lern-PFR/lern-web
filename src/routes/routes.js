/* eslint-disable react/jsx-filename-extension */
/* eslint-disable react/display-name */
import { PostSignupPage, SignInPage, SignupPage } from 'pages/auth';
import { ConceptCreationPage, ConceptDetailsPage, ConceptEditionPage } from 'pages/concepts';
import { SubjectCreationPage, SubjectDetailsPage, SubjectEditionPage, SubjectListPage, SubjectStructurePage } from 'pages/subjects';
import { ModuleCreationPage, ModuleDetailsPage, ModuleEditionPage } from 'pages/modules';
import { Homepage } from 'pages/home';
import { LessonCreationPage } from 'pages/lessons';
import keys from './keys';

/**
 * Routes to render
 * For authentication purposes, each route can have `allowAnonymous` and `allowAuthenticated` values
 * Default values are :
 *  - `allowAnonymous`		: false
 *  - `allowAuthenticated`	: true
 */
export default [
	{
		path: keys.app.default,
		exact: true,
		allowAnonymous: true,
		component: Homepage,
	},
	{
		path: keys.about.default,
		exact: true,
		allowAnonymous: true,
		allowAuthenticated: true,
		component: () => <p>About</p>,
	},
	{
		path: keys.contactUs.default,
		exact: true,
		allowAnonymous: true,
		allowAuthenticated: true,
		component: () => <p>Contact us</p>,
	},
	// Home
	{
		path: keys.home.default,
		exact: true,
		allowAnonymous: true,
		allowAuthenticated: true,
		component: Homepage,
	},
	// Auth
	{
		path: keys.auth.default,
		routes: [
			{
				path: keys.auth.login,
				allowAnonymous: true,
				allowAuthenticated: false,
				exact: true,
				component: SignInPage,
			},
			{
				path: keys.auth.logout,
				exact: true,
				component: () => <p>Logout</p>,
			},
			{
				path: keys.auth.forgottenPassword,
				allowAnonymous: true,
				allowAuthenticated: false,
				exact: true,
				component: () => <p>Forgot password</p>,
			},
			{
				path: keys.auth.signup,
				allowAnonymous: true,
				allowAuthenticated: false,
				exact: true,
				component: SignupPage,
			},
			{
				path: keys.auth.postSignup,
				allowAnonymous: true,
				allowAuthenticated: false,
				exact: true,
				component: PostSignupPage,
			},
		],
	},
	// Subjects
	{
		path: keys.subjects.default,
		routes: [
			{
				path: keys.subjects.list,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: SubjectListPage,
			},
			{
				path: keys.subjects.subjectStructure,
				allowAnonymous: true,
				allowAuthenticated: true,
				exact: true,
				component: SubjectStructurePage,
			},
			{
				path: keys.subjects.subjectCreation,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: SubjectCreationPage,
			},
			{
				path: keys.subjects.subjectEdition,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: SubjectEditionPage,
			},
			{
				path: keys.subjects.moduleCreation,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ModuleCreationPage,
			},
			{
				path: keys.subjects.subjectDetails,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: SubjectDetailsPage,
			},
		],
	},
	// Modules
	{
		path: keys.modules.default,
		routes: [
			{
				path: keys.modules.moduleEdition,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ModuleEditionPage,
			},
			{
				path: keys.modules.conceptCreation,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ConceptCreationPage,
			},
			{
				path: keys.modules.moduleDetails,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ModuleDetailsPage,
			},
		],
	},
	// Concepts
	{
		path: keys.concepts.default,
		routes: [
			{
				path: keys.concepts.conceptEdition,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ConceptEditionPage,
			},
			{
				path: keys.concepts.conceptDetails,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ConceptDetailsPage,
			},
			{
				path: keys.concepts.lessonCreation,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: LessonCreationPage,
			},
			{
				path: keys.concepts.lessonDetails,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ConceptDetailsPage,
			},
		],
	},
	// Lessons
	{
		path: keys.lessons.default,
		routes: [
			{
				path: keys.lessons.lessonEdition,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: () => <p>Lesson edition</p>,
			},
		],
	},
];
