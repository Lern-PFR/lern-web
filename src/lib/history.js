import { createBrowserHistory } from 'history';
import { clearAlerts } from '../redux/actions/alerts';
import store from '../redux/store/configureStore';

const { dispatch } = store;

// Init applicationPathname
const history = createBrowserHistory();

history.listen(() => {
  dispatch(clearAlerts());
});

export default history;
