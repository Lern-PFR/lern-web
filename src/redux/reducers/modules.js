import { ActionTypes } from 'redux/actions/modules';

const initialState = {
	isLoading: false,
	items: [],
	totalCount: 0,
};

export default (state = initialState, { type, payload }) => {
	switch (type) {
		case ActionTypes.FETCH_MODULE_REQUEST:
		case ActionTypes.FETCH_MODULE_LIST_REQUEST:
		case ActionTypes.CREATE_MODULE_REQUEST:
		case ActionTypes.UPDATE_MODULE_REQUEST:
		case ActionTypes.DELETE_MODULE_REQUEST:
			return {
				...state,
				isLoading: true,
			};
		case ActionTypes.FETCH_MODULE_FAILURE:
		case ActionTypes.FETCH_MODULE_LIST_FAILURE:
		case ActionTypes.CREATE_MODULE_FAILURE:
		case ActionTypes.UPDATE_MODULE_FAILURE:
		case ActionTypes.DELETE_MODULE_FAILURE:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.CREATE_MODULE_SUCCESS:
		case ActionTypes.UPDATE_MODULE_SUCCESS:
		case ActionTypes.DELETE_MODULE_SUCCESS:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.FETCH_MODULE_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [payload.module],
				totalCount: 1,
			};
		case ActionTypes.FETCH_MODULE_LIST_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [...payload.modules],
				totalCount: payload.totalCount,
			};
		case ActionTypes.CLEAR_MODULE_LIST:
			return {
				...state,
				items: [],
				totalCount: 0,
			};
		default:
			return state;
	}
};
