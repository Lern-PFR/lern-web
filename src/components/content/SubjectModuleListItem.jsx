import { useCallback, useMemo } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { generatePath } from 'react-router';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';

import routes from 'routes';
import { deleteModule } from 'redux/actions/modules';
import { StyledDiv, StyledListItem } from 'components/shared/styledElements';
import { BodyCopy, Brevier } from 'components/shared/typography';
import { OutlinedButton, OutlinedLinkButton } from 'components/shared/buttons';
import Link from 'components/shared/navigation/Link';

import { innerContentListItem, innerContentListItemActionsRow, moduleCard } from 'theme/contentEditionCommon/innerContentListTheme';
import { history } from 'routes/components/RouterProvider';

const ListCard = styled(StyledListItem)(
	{
		...innerContentListItem,
		'&, & *': {
			cursor: 'pointer',
		},
	}
);

/**
 * @name SubjectModuleListItem
 * @description A component used to display a subject's module in the subject edition page.
 *
 * @author TimothÃ©e Simon-Franza
 *
 */
const SubjectModuleListItem = ({ description, id, title }) => {
	const dispatch = useDispatch();
	const { t } = useTranslation();

	const editPageRedirectionLink = useMemo(() => generatePath(routes.modules.moduleEdition, { moduleId: id }), [id]);

	const onCardClick = useCallback(() => {
		history.push(editPageRedirectionLink);
	}, [editPageRedirectionLink]);

	const onDeleteModuleBtnClick = useCallback(() => {
		dispatch(deleteModule(id));
	}, [dispatch, id]);

	return (
		<ListCard key={id} onClick={onCardClick} {...moduleCard}>
			<BodyCopy margin="0">{title}</BodyCopy>
			<Brevier margin="0">{description}</Brevier>
			<StyledDiv {...innerContentListItemActionsRow}>
				<OutlinedLinkButton>
					<Link to={editPageRedirectionLink}>{t('subjects.edition.modules_list.links.edit')}</Link>
				</OutlinedLinkButton>
				<OutlinedButton onClick={onDeleteModuleBtnClick}>
					{t('subjects.edition.modules_list.links.delete')}
				</OutlinedButton>
			</StyledDiv>
		</ListCard>
	);
};

SubjectModuleListItem.propTypes = {
	description: PropTypes.string.isRequired,
	id: PropTypes.string.isRequired,
	title: PropTypes.string.isRequired,
};

export default SubjectModuleListItem;
