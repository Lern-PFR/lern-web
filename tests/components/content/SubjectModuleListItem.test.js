import { shallow } from 'enzyme';
import { act, fireEvent, render, screen } from '@testing-library/react';

import { Router, StaticRouter } from 'react-router';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import SubjectModuleListItem from 'components/content/SubjectModuleListItem';

import { ActionTypes } from 'redux/actions/modules';
import { createMemoryHistory } from 'history';
import routes from 'routes';

describe('Subject edition page - module list', () => {
	const mockStore = configureMockStore([thunk]);
	const store = mockStore({});
	const mockedModule = { id: 'dummy_module_id_1', title: 'dummy_module_title_1', description: 'dummy_module_desc_1' };

	afterEach(() => {
		jest.restoreAllMocks();
		jest.resetAllMocks();
	});

	describe('snapshots', () => {
		it('should match previous snapshot', () => {
			const sut = (
				<Provider store={store}>
					<SubjectModuleListItem {...mockedModule} />
				</Provider>
			);
			const wrapper = shallow(sut);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('onClick events', () => {
		it('should dispatch a module deletion request action upon click on the module card\'s delete button.', () => {
			render(
				<StaticRouter>
					<Provider store={store}>
						<SubjectModuleListItem {...mockedModule} />
					</Provider>
				</StaticRouter>
			);

			act(() => {
				fireEvent.click(screen.getByText('subjects.edition.modules_list.links.delete'));
			});

			expect(store.getActions()).toEqual([{ type: ActionTypes.DELETE_MODULE_REQUEST }]);
		});

		it('should redirect to the module\'s edition page url upon click on the card\'s edit button.', () => {
			const history = createMemoryHistory();

			render(
				<Router history={history}>
					<Provider store={store}>
						<SubjectModuleListItem {...mockedModule} />
					</Provider>
				</Router>
			);

			act(() => {
				fireEvent.click(screen.getByText('subjects.edition.modules_list.links.edit'));
			});

			expect(history.location.pathname).toEqual(routes.modules.moduleEdition.replace(':moduleId', mockedModule.id));
		});
	});
});
