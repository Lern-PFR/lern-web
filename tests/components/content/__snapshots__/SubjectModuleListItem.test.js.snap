// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subject edition page - module list snapshots should match previous snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <SubjectModuleListItem
    description="dummy_module_desc_1"
    id="dummy_module_id_1"
    title="dummy_module_title_1"
  />
</ContextProvider>
`;
